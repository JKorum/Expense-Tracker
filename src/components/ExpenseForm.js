import React from 'react'
import moment from 'moment'
import uuid from 'uuid'

import { SingleDatePicker } from 'react-dates'

class ExpenseForm extends React.Component {
	constructor(props) {
		super(props)
		this.state = {
			description: props.expense? props.expense.description : ``,
			note: props.expense? props.expense.note : ``,
			amount: props.expense? props.expense.amount.toString() : ``,
			createdAt: props.expense? moment(props.expense.createdAt) : moment(),
			calendarFocused: false,
			error: undefined,
			button: props.buttonName
		}

		this.handleDescriptionChange = this.handleDescriptionChange.bind(this)
		this.handleNoteChange = this.handleNoteChange.bind(this)
		this.handleAmountChange = this.handleAmountChange.bind(this)
		this.handleDateChange = this.handleDateChange.bind(this)
		this.handleFocusChange = this.handleFocusChange.bind(this)
		this.handleSubmit = this.handleSubmit.bind(this)
	}

	handleDescriptionChange(e) {
		const description = e.target.value
		this.setState(() => ({ description }))
	}

	handleNoteChange(e) {
		const note = e.target.value
		this.setState(() => ({ note }))
	}

	handleAmountChange(e) {
		const amount = e.target.value
		if (!amount || /^\d+(\.\d{0,2})?$/.test(amount)) { 
			this.setState(() => ({ amount }))
		}		
	}

	handleDateChange(calendarDate) {
		if (calendarDate) {
			this.setState(() => ({ createdAt: calendarDate }))
		}		
	}

	handleFocusChange({ focused }) {
		this.setState(() => ({ calendarFocused: focused }))
	}

	handleSubmit(e) {
		e.preventDefault()

		if (!this.state.description || !this.state.amount) {
			this.setState(() => ({ error: `Please provide description and amount` }))
		} else {
				this.setState(() => ({ error: undefined }))
				this.props.onSubmit({
					description: this.state.description,
					amount: +this.state.amount,
					createdAt: this.state.createdAt.valueOf(), // returns timestamp from moment obj
					note: this.state.note
				})
		}
	}

	render() {		
		return (				
			<form className="form" onSubmit={this.handleSubmit}>
				{this.state.error && <p className="form__error">{this.state.error}</p>}
				<input 
					type="text" 
					className="text-input"
					placeholder="description" 
					autoFocus 
					value={this.state.description}
					onChange={this.handleDescriptionChange}
				/>

				<input 
					type="text" 
					className="text-input"
					value={this.state.amount}
					onChange={this.handleAmountChange}
					placeholder="amount" 
				/>

				<SingleDatePicker
				  date={this.state.createdAt} 
				  onDateChange={this.handleDateChange}
				  focused={this.state.calendarFocused}
				  onFocusChange={this.handleFocusChange}
				  id={uuid()}
				  numberOfMonths={1}		
				  isOutsideRange={() => false} // make days in a past clickable			  
				/>

				<textarea 
					className="textarea"
					value={this.state.note}
					onChange={this.handleNoteChange}
					placeholder="add a note for your expense"
				>
				</textarea>
				<div>
					<button 
						type="submit"
						className="button"
						disabled={this.props.buttonName === `edit expense` && this.props.expense === undefined}
					>{this.state.button}</button>
				</div>				
			</form>			
		)
	}
}

export default ExpenseForm