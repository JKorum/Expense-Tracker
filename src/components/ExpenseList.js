import React from 'react'
import { connect } from 'react-redux'

import getVisibleExpenses from '../selectors/expenses'
import ExpenseListItem from './ExpenseListItem'

//component declaration props expects to be connected to store
export const ExpenseList = ({ expenses }) => (
	<div>
		{
			expenses.length === 0? (
				<p>No expenses</p>
			) : (
				expenses.map(expense => <ExpenseListItem key={expense.id} {...expense} />)
			) 
		}		
	</div>
)

// what store properties connected component should have access to
const mapStateToProps = (state) => ({
	expenses: getVisibleExpenses(state.expenses, state.filters)
})

//connect()() returns connected component 
export default connect(mapStateToProps)(ExpenseList)