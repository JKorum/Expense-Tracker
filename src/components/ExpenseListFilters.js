import React from 'react'
import { connect } from 'react-redux'
import uuid from 'uuid'
import { 
	setTextFilter, 
	sortByAmount, 
	sortByDate,
	setStartDate, 
	setEndDate 
} from '../actions/filters'

import { DateRangePicker } from 'react-dates'

//component declaration props expects to be connected to store
export class ExpenseListFilters extends React.Component {
	constructor(props) {
		super(props)
		this.state = {
			calendarFocused: null
		}

		this.handleCalendarDatesChange = this.handleCalendarDatesChange.bind(this)
		this.handleCalendarFocusChange = this.handleCalendarFocusChange.bind(this)
		this.handleTextChange = this.handleTextChange.bind(this)
		this.handleSortChange = this.handleSortChange.bind(this)
	}

	handleCalendarDatesChange({ startDate, endDate }) {		
		this.props.setStartDate(startDate)
		this.props.setEndDate(endDate)		
	}

	handleCalendarFocusChange(calendarFocused) {
		this.setState(() => ({ calendarFocused }))
	}

	handleTextChange(e) {		
		this.props.setTextFilter(e.target.value)
	}

	handleSortChange(e) {
		e.target.value === `date`? this.props.sortByDate() : this.props.sortByAmount() 
	}

	render() {
		const { filters, dispatch } = this.props

		return (
			<div>
				<input type="text" value={filters.text} onChange={this.handleTextChange} />
				<select value={filters.sortBy} onChange={this.handleSortChange}>
					<option value="date">Date</option>
					<option value="amount">Amount</option>
				</select>
				<DateRangePicker
				  startDate={filters.startDate} 
				  startDateId={uuid()} 
				  endDate={filters.endDate} 
				  endDateId={uuid()} 
				  onDatesChange={this.handleCalendarDatesChange}
				  focusedInput={this.state.calendarFocused}
				  onFocusChange={this.handleCalendarFocusChange}
				 	numberOfMonths={1}		
					isOutsideRange={() => false} // make days in a past clickable	
					showClearDates={true}		
				/>
			</div>
		)
	}
}

// what store properties connected component should have access to
const mapStateToProps = (state) => ({
	filters: state.filters
})

const mapDispatchToProps = (dispatch) => ({
	setStartDate: (startDate) => dispatch(setStartDate(startDate)),
	setEndDate: (endDate) => dispatch(setEndDate(endDate)),
	setTextFilter: (text) => dispatch(setTextFilter(text)),
	sortByDate: () => dispatch(sortByDate()),
	sortByAmount: () => dispatch(sortByAmount())
})

//connect()() returns connected component 
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)
