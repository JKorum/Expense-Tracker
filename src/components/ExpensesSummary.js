import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import numeral from 'numeral'

import getVisibleExpenses from '../selectors/expenses'
import selectExpensesTotal from '../selectors/selectExpensesTotal'

export const ExpensesSummary = ({ counter, total }) => (
	<div className="page-header">
		<div className="content-container">
			<h3 className="page-header__title">
				{`Viewing ${counter === 1 
					? `one expense` 
					: `${counter} expenses`} totaling ${numeral(total).format(`$0,0.00`)}`}
			</h3>
			<div className="page-header__actions">
				<Link className="button" to="/create">Add expense</Link>
			</div>
		</div>		
	</div>
)

const mapStateToProps = ({ expenses, filters }) => {
	const visibleExpenses = getVisibleExpenses(expenses, filters) 
	return {
		counter: visibleExpenses.length, 
		total: selectExpensesTotal(visibleExpenses) 
	}
}

export default connect(mapStateToProps)(ExpensesSummary)