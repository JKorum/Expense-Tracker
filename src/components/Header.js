import React from 'react'
import { connect } from 'react-redux'
import { NavLink } from 'react-router-dom'
import { startSignout } from '../actions/auth'

export const Header = ({ startSignout }) => (
	<div>
		<h1>Expense tracker</h1>		
		<NavLink to="/dashboard" activeClassName="is-active">Dashboard</NavLink>
		<NavLink to="/create" activeClassName="is-active">Create expense</NavLink>				
		<button onClick={startSignout}>Logout</button>		
	</div>
)

const mapDispatchToProps = (dispatch) => ({
	startSignout: () => dispatch(startSignout())
})

export default connect(undefined, mapDispatchToProps)(Header)