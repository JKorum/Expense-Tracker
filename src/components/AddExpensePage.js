import React from 'react'
import { connect } from 'react-redux'

import ExpenseForm from './ExpenseForm'
import { addExpense } from '../actions/expenses'

export class AddExpensePage extends React.Component {
	constructor(props) {
		super(props)
		this.handleSubmit = this.handleSubmit.bind(this)
	}

	handleSubmit(expense) {
		this.props.addExpense(expense)
		this.props.history.push(`/`)
	}

	render() {
		return (
			<div>
				<h1>Add Expense</h1>
				<ExpenseForm 
					buttonName="add expense"
					onSubmit={this.handleSubmit} 
				/>
			</div>
		)
	}
}

const mapDispatchToProps = (dispatch) => ({
	addExpense: (expense) => dispatch(addExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage)