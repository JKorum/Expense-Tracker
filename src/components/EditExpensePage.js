import React from 'react'
import { connect } from 'react-redux'

import ExpenseForm from './ExpenseForm'
import { startEditExpense, startDeleteExpense } from '../actions/expenses'

export class EditExpensePage extends React.Component {
	constructor(props) {
		super(props)
		this.handleSubmit = this.handleSubmit.bind(this)
		this.handleRemoveExpense = this.handleRemoveExpense.bind(this)
	}

	handleSubmit(expense) {		
		this.props.startEditExpense(this.props.expense.id, expense)
		this.props.history.push(`/`)
	}

	handleRemoveExpense(e){		
		this.props.startDeleteExpense(this.props.expense.id)
		this.props.history.push(`/`)
	}

	render() {
		return (
			<div>
				<div className="page-header">
					<div className="content-container">
						<h3 className="page-header__title">Edit Expense</h3>
					</div>
				</div>
				<div className="content-container">
					<ExpenseForm 
						buttonName="edit expense"
						expense={this.props.expense}
						onSubmit={this.handleSubmit}
					/>
					<button 
						type="button"
						className="button button--secondary" 
						onClick={this.handleRemoveExpense} 
						disabled={!this.props.expense}
					>Remove expense</button>
				</div>				
			</div>
		)
	}
}

const mapStateToProps = (state, ownProps) => ({
	expense: state.expenses.find(expense => expense.id === ownProps.match.params.id)
})

const mapDispatchToProps = (dispatch) => ({
	startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
	startDeleteExpense: id => dispatch(startDeleteExpense(id))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)