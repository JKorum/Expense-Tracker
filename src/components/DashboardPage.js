import React from 'react'
import { connect } from 'react-redux'
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
import ExpensesSummary from './ExpensesSummary'
import getVisibleExpenses from '../selectors/expenses'

export const DashboardPage = ({ counter }) => (
	<div>
		{ counter > 0 && <ExpensesSummary /> }
		<ExpenseListFilters />
		<ExpenseList />
	</div>
)

const mapStateToProps = ({ expenses, filters }) => ({
	counter: getVisibleExpenses(expenses, filters).length
})

export default connect(mapStateToProps)(DashboardPage)