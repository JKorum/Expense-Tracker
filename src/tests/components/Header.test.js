import React from 'react'
import { shallow } from 'enzyme'
import { Header } from '../../components/Header'

test(`should render Header correctly`, () => {
	const wrapper = shallow(<Header startSignout={() => {}}/>)
	expect(wrapper).toMatchSnapshot()
})

test(`should call startSignout on button click`, () => {
	const spyStartSignout = jest.fn()
	const wrapper = shallow(<Header startSignout={spyStartSignout}/>)
	wrapper.find('button').simulate('click')
	expect(spyStartSignout).toHaveBeenCalled()
})