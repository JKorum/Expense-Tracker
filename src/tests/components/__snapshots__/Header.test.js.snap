// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<div>
  <h1>
    Expense tracker
  </h1>
  <NavLink
    activeClassName="is-active"
    to="/dashboard"
  >
    Dashboard
  </NavLink>
  <NavLink
    activeClassName="is-active"
    to="/create"
  >
    Create expense
  </NavLink>
  <button
    onClick={[Function]}
  >
    Logout
  </button>
</div>
`;
